Amethyst/Layout/ReflowOperation.swift

```diff
@@ -62,6 +62,177 @@ open class ReflowOperation: Operation {
+            return true
+        }
+        if window_frame.origin.y.isNaN {
+            return true
+        }
+        if window_frame.width <= 0.0 {
+            return true
+        }
+        if window_frame.width.isInfinite {
+            return true
+        }
+        if window_frame.width.isNaN {
+            return true
+        }
+        if window_frame.height <= 0.0 {
+            return true
+        }
+        if window_frame.height.isInfinite {
+            return true
+        }
+        if window_frame.height.isNaN {
+            return true
+        }
+        return false
+    }
+
+    private func categorizeWindows(candidateFrames: [CGRect]) -> ([SIWindow], [SIWindow]) {
+
+        var displaced: [SIWindow] = []
+        var aligned: [SIWindow] = []
+        for window in self.windows {
+            //Sometimes the windows become degenerate for some reason, ignore them
+            if !isDegenerateWindow(window:window) {
+                var min_distance = FLT_MAX
+                let window_frame = window.frame()
+                let windows_center_x = window_frame.origin.x + (window_frame.width/2.0)
+                let windows_center_y = window_frame.origin.y + (window_frame.height/2.0)
+                for frame in candidateFrames {
+                    let frame_center_x = frame.origin.x + (frame.width/2.0)
+                    let frame_center_y = frame.origin.y + (frame.height/2.0)
+                    let distance_x = (frame_center_x - windows_center_x)
+                    let distance_y = (frame_center_y - windows_center_y)
+                    let distance = sqrt(distance_x*distance_x + distance_y*distance_y)
+                    if Float(distance) < min_distance {
+                        min_distance = Float(distance)
+                    }
+                }
+                if min_distance < 1.0 {
+                    aligned.append(window)
+                }
+                else {
+                    displaced.append(window)
+                }
+            }
+        }
+        return (displaced, aligned)
+
+    }
+
+    open func assignWindowsToFramesBasedOnDistance(candidateFrames: [CGRect]) {
+
+        var candidateFramesMutable = candidateFrames
+        //Find those windows that are not perfectly on top of a frame to assign first, because those are the ones that have been newly created or manually moved
+        let (displacedWindows, alignedWindows) = categorizeWindows(candidateFrames: candidateFrames)
+
+        //Now preferentially assign displaced windows a location
+        let screenFrame = adjustedFrameForLayout(screen)
+        let focusedWindow = SIWindow.focused()
+        var frameAssignments: [FrameAssignment] = []
+        for window in displacedWindows {
+            var min_distance = FLT_MAX
+            var min_index = -1
+
+            //Sometimes the windows become degenerate for some reason
+            if !isDegenerateWindow(window:window) {
```

Comment:
This degenerate window thing is worrying me. It feels like there's a deeper problem there that could be the source of existing bugs elsewhere.